coredb-operator/src/fixtures.rs

```diff
@@ -0,0 +1,150 @@
+        d.spec.replicas = 1;
+        d
+    }
+
+    /// Modify a coredb to have the expected finalizer
+    pub fn finalized(mut self) -> Self {
+        self.finalizers_mut().push(COREDB_FINALIZER.to_string());
+        self
+    }
+
+    /// Modify a coredb to have an expected status
+    pub fn with_status(mut self, status: CoreDBStatus) -> Self {
+        self.status = Some(status);
+        self
+    }
+}
+pub struct ApiServerVerifier(Handle<Request<Body>, Response<Body>>);
+
+/// Create a responder + verifier object that deals with the main reconcile scenarios
+///
+/// 1. objects without finalizers will get a finalizer applied (and not call the apply loop)
+/// 2. finalized objects will run through the apply loop
+/// 3. finalized objects "with errors" (i.e. the "illegal" object) will short circuit the apply loop
+/// 4. objects with a deletion timestamp will run the cleanup loop (which will send an event)
+impl ApiServerVerifier {
+    pub fn handle_finalizer_creation(self, coredb_: &CoreDB) -> JoinHandle<()> {
+        let handle = self.0;
+        let coredb = coredb_.clone();
+        tokio::spawn(async move {
+            pin_mut!(handle);
+            let (request, send) = handle.next_request().await.expect("service not called");
+            // We expect a json patch to the specified coredb adding our finalizer
+            assert_eq!(request.method(), http::Method::PATCH);
+            assert_eq!(
+                request.uri().to_string(),
+                format!("/apis/kube.rs/v1/namespaces/testns/coredbs/{}?", coredb.name_any())
+            );
+            let expected_patch = serde_json::json!([
+                { "op": "test", "path": "/metadata/finalizers", "value": null },
+                { "op": "add", "path": "/metadata/finalizers", "value": vec![COREDB_FINALIZER] }
+            ]);
+            let req_body = to_bytes(request.into_body()).await.unwrap();
+            let runtime_patch: serde_json::Value =
+                serde_json::from_slice(&req_body).expect("valid coredb from runtime");
+            assert_json_include!(actual: runtime_patch, expected: expected_patch);
+
+            let response = serde_json::to_vec(&coredb.finalized()).unwrap(); // respond as the apiserver would have
+            send.send_response(Response::builder().body(Body::from(response)).unwrap());
+        })
+    }
+
+    pub fn handle_event_publish(self) -> JoinHandle<()> {
+        let handle = self.0;
+        tokio::spawn(async move {
+            pin_mut!(handle);
+            let (request, send) = handle.next_request().await.expect("service not called");
+            assert_eq!(request.method(), http::Method::POST);
+            assert_eq!(
+                request.uri().to_string(),
+                format!("/apis/events.k8s.io/v1/namespaces/testns/events?")
+            );
+            // pass the event straight through
+            send.send_response(Response::builder().body(request.into_body()).unwrap());
+        })
+    }
```

Comment:
I am reviewing this part, I think that in the example controller, they are publishing events, but not in our case. So I think I should delete this part.

https://github.com/kube-rs/controller-rs/blob/e20529654060d8bbd067be59a828f6f6f7881cba/src/controller.rs#L94-L106